In [54]:
# load libraries
import pandas as pd
import datetime
import matplotlib.pyplot as plt

### Initial data check

In [55]:
# import data
df_campfin = pd.read_csv('Campaign_Finance_20260204.csv')

In [56]:
df_campfin.head()

ID  Record ID  Report ID  \
0  FR12995       2995       2995   
1  FR12999       2999       2999   
2  FR12998       2998       2998   
3  FR12997       2997       2997   
4  FR12992       2992       2992   

                                           File Link First Name Last Name  \
0  https://campfin.dallascityhall.com/FinalReport...    Carolyn    Arnold   
1  https://campfin.dallascityhall.com/FinalReport...      Zarin    Gracey   
2  https://campfin.dallascityhall.com/FinalReport...      Jaime  Resendez   
3  https://campfin.dallascityhall.com/FinalReport...   Monica R    Alonzo   
4  https://campfin.dallascityhall.com/FinalReport...      Jesus    Moreno   

  Business Name           Contact Type                   Record Type  Amount  \
0           NaN  Candidate / Committee  January 15: Semi-Annual 2026     0.0   
1           NaN  Candidate / Committee  January 15: Semi-Annual 2026     0.0   
2           NaN  Candidate / Committee  January 15: Semi-Annual 2026     0.0   
3           NaN  Candidate / Committee  January 15: Semi-Annual 2026     0.0   
4           NaN  Candidate / Committee  January 15: Semi-Annual 2026     0.0   

  Schedule Type   Candidate Name           Election Date  \
0        Report   Carolyn Arnold                     NaN   
1        Report     Zarin Gracey                     NaN   
2        Report   Jaime Resendez  05/03/2025 12:00:00 AM   
3        Report  Monica R Alonzo  05/03/2025 12:00:00 AM   
4        Report     Jesus Moreno  11/01/2027 12:00:00 AM   

         Transaction Date                 Street    City State Zipcode  \
0  01/15/2026 05:54:31 PM   835 Timber Dell Lane  Dallas    TX   75232   
1  01/15/2026 04:48:27 PM          PO Box 763173  Dallas    TX   75376   
2  01/15/2026 04:30:46 PM  1458 Mission Hills Ln  Dallas    TX   75217   
3  01/15/2026 04:21:23 PM            PO Box 4126  Dallas    TX   75208   
4  01/15/2026 04:21:19 PM         4809 Victor St  Dallas    TX   75246   

                                        Geo Location  
0  835 Timber Dell Lane Dallas TX 75232 (32.67779...  
1                            763173 Dallas TX  75376  
2  1458 Mission Hills Ln Dallas TX 75217 (32.7379...  
3        4126 Dallas TX  75208 (33.11273, -96.82308)  
4  4809 Victor St Dallas TX 75246 (32.79619, -96....

In [57]:
df_campfin.tail()

ID  Record ID  Report ID  \
9945  F45852      45852       2735   
9946  F46381      46381       2754   
9947  F46358      46358       2746   
9948  F45906      45906       2735   
9949  F46362      46362       2756   

                                              File Link First Name Last Name  \
9945  https://campfin.dallascityhall.com/FinalReport...        NaN       NaN   
9946  https://campfin.dallascityhall.com/FinalReport...        NaN       NaN   
9947  https://campfin.dallascityhall.com/FinalReport...        NaN       NaN   
9948  https://campfin.dallascityhall.com/FinalReport...        NaN       NaN   
9949  https://campfin.dallascityhall.com/FinalReport...        NaN       NaN   

           Business Name Contact Type                Record Type  Amount  \
9945           Sams Club  Expenditure  July 15: Semi-Annual 2024   41.77   
9946        Square Space  Expenditure  July 15: Semi-Annual 2024    6.50   
9947           MAILCHIMP  Expenditure  July 15: Semi-Annual 2024  117.26   
9948  Jefferson Monument  Expenditure  July 15: Semi-Annual 2024  350.00   
9949  SiteGround Hosting  Expenditure  July 15: Semi-Annual 2024   19.99   

               Schedule Type  Candidate Name           Election Date  \
9945  Political Expenditures  Tennell Atkins                     NaN   
9946  Political Expenditures    Jesus Moreno  05/06/2023 12:00:00 AM   
9947  Political Expenditures  Paula Blackmon  05/03/2025 12:00:00 AM   
9948  Political Expenditures  Tennell Atkins                     NaN   
9949  Political Expenditures  Jaime Resendez  05/06/2023 12:00:00 AM   

            Transaction Date               Street           City State  \
9945  01/02/2024 12:00:00 AM  2900 W Wheatland Rd         Dallas    TX   
9946  01/02/2024 12:00:00 AM         1455 Main St  San Francisco    CA   
9947  01/02/2024 12:00:00 AM  405 N Angier Ave NE        Atlanta    GA   
9948  01/02/2024 12:00:00 AM        351 Jefferson         Dallas    TX   
9949  01/01/2024 12:00:00 AM    14455 n Hayden Rd     Scottsdale    AR   

     Zipcode                                       Geo Location  
9945   75237  2900 W Wheatland Rd Dallas TX 75237 (32.64542,...  
9946   94104                1455 Main St San Francisco CA 94104  
9947   30308  405 N Angier Ave NE Atlanta GA 30308 (33.76966...  
9948   75208  351 Jefferson Dallas TX 75208 (32.74318, -96.8...  
9949   85260  14455 n Hayden Rd Scottsdale AR 85260 (33.6185...

In [58]:
# how many rows in the dataset?
len(df_campfin)

9950

In [59]:
# each row has a unique ID
df_campfin['ID'].nunique()

9950

In [60]:
# nas in the first name, last name and business name columns make sense, since donors will either be individuals or businesses
df_campfin.isnull().sum()

ID                     0
Record ID              0
Report ID              0
File Link              0
First Name          4005
Last Name           4006
Business Name       5682
Contact Type           0
Record Type            0
Amount                 0
Schedule Type          0
Candidate Name         0
Election Date       1050
Transaction Date       0
Street                 0
City                   0
State                  0
Zipcode                0
Geo Location           0
dtype: int64

In [61]:
# check data types -- we want to convert the transaction date and election date to datetime objects
df_campfin.dtypes

ID                      str
Record ID             int64
Report ID             int64
File Link               str
First Name              str
Last Name               str
Business Name           str
Contact Type            str
Record Type             str
Amount              float64
Schedule Type           str
Candidate Name          str
Election Date           str
Transaction Date        str
Street                  str
City                    str
State                   str
Zipcode                 str
Geo Location            str
dtype: object

In [62]:
# change date columns to datetime
df_campfin['Transaction Date'] = pd.to_datetime(df_campfin['Transaction Date'], errors='coerce')
df_campfin['Election Date'] = pd.to_datetime(df_campfin['Election Date'], errors='coerce')


/var/folders/sv/fm7yfyrs7317lzyxf_8p39800000gn/T/ipykernel_61275/1085802198.py:2: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df_campfin['Transaction Date'] = pd.to_datetime(df_campfin['Transaction Date'], errors='coerce')
/var/folders/sv/fm7yfyrs7317lzyxf_8p39800000gn/T/ipykernel_61275/1085802198.py:3: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df_campfin['Election Date'] = pd.to_datetime(df_campfin['Election Date'], errors='coerce')


In [63]:
df_campfin.dtypes

ID                             str
Record ID                    int64
Report ID                    int64
File Link                      str
First Name                     str
Last Name                      str
Business Name                  str
Contact Type                   str
Record Type                    str
Amount                     float64
Schedule Type                  str
Candidate Name                 str
Election Date       datetime64[us]
Transaction Date    datetime64[us]
Street                         str
City                           str
State                          str
Zipcode                        str
Geo Location                   str
dtype: object

In [64]:
# earliest and latest election dates - 2019 election (Dallas mayoral?) to 2024 election
print(df_campfin['Election Date'].min())
print(df_campfin['Election Date'].max())

2019-05-04 00:00:00
2027-11-07 00:00:00


In [77]:
# how many unique elections, and what are they
df_campfin['Election Date'].nunique()
df_campfin['Election Date'].unique()

<DatetimeArray>
[                'NaT', '2025-05-03 00:00:00', '2027-11-01 00:00:00',
 '2027-11-07 00:00:00', '2026-05-02 00:00:00', '2023-05-06 00:00:00',
 '2027-11-02 00:00:00', '2025-06-07 00:00:00', '2027-05-01 00:00:00',
 '2025-05-05 00:00:00', '2025-05-10 00:00:00', '2025-04-30 00:00:00',
 '2021-05-01 00:00:00', '2024-05-03 00:00:00', '2019-05-04 00:00:00',
 '2024-05-10 00:00:00', '2025-06-01 00:00:00']
Length: 17, dtype: datetime64[us]

In [ ]:
# basic statistics 
df_campfin.describe()

Record ID    Report ID        Amount               Election Date  \
count   9950.000000  9950.000000   9950.000000                        8900   
mean   69462.849648  2863.928643    587.842293  2025-05-18 14:07:29.797752   
min      105.000000  2732.000000      0.000000         2019-05-04 00:00:00   
25%    48057.250000  2810.000000     54.610000         2025-05-03 00:00:00   
50%    88368.000000  2846.000000    173.115000         2025-05-03 00:00:00   
75%    92128.750000  2938.000000    500.000000         2025-05-03 00:00:00   
max    95319.000000  2999.000000  76000.000000         2027-11-07 00:00:00   
std    25634.987485    75.420379   2037.281922                         NaN   

                 Transaction Date  
count                        9950  
mean   2025-02-14 09:44:05.881306  
min           2024-01-01 00:00:00  
25%           2025-01-03 00:00:00  
50%           2025-03-18 05:59:15  
75%           2025-04-26 00:00:00  
max           2026-01-15 17:54:31  
std                           NaN

The average contribution is about $588, and contributions range from $0(?) to $76,000. But the median is much lower 

In [66]:
# the minimum is zero dollars, which is a little weird
df_campfin[df_campfin['Amount'] == 0]

ID  Record ID  Report ID  \
0     FR12995       2995       2995   
1     FR12999       2999       2999   
2     FR12998       2998       2998   
3     FR12997       2997       2997   
4     FR12992       2992       2992   
...       ...        ...        ...   
8994  FR12733       2733       2733   
9002  FR12739       2739       2739   
9010  FR12737       2737       2737   
9014  FR12736       2736       2736   
9022  FR12734       2734       2734   

                                              File Link First Name  \
0     https://campfin.dallascityhall.com/FinalReport...    Carolyn   
1     https://campfin.dallascityhall.com/FinalReport...      Zarin   
2     https://campfin.dallascityhall.com/FinalReport...      Jaime   
3     https://campfin.dallascityhall.com/FinalReport...   Monica R   
4     https://campfin.dallascityhall.com/FinalReport...      Jesus   
...                                                 ...        ...   
8994  https://campfin.dallascityhall.com/FinalReport...      Kathy   
9002  https://campfin.dallascityhall.com/FinalReport...      Sandy   
9010  https://campfin.dallascityhall.com/FinalReport...       Cara   
9014  https://campfin.dallascityhall.com/FinalReport...     Jaynie   
9022  https://campfin.dallascityhall.com/FinalReport...       Paul   

       Last Name Business Name           Contact Type  \
0         Arnold           NaN  Candidate / Committee   
1         Gracey           NaN  Candidate / Committee   
2       Resendez           NaN  Candidate / Committee   
3         Alonzo           NaN  Candidate / Committee   
4         Moreno           NaN  Candidate / Committee   
...          ...           ...                    ...   
8994     Stewart           NaN  Candidate / Committee   
9002     Greyson           NaN  Candidate / Committee   
9010  Mendelsohn           NaN  Candidate / Committee   
9014     Schultz           NaN  Candidate / Committee   
9022      Ridley           NaN  Candidate / Committee   

                       Record Type  Amount Schedule Type   Candidate Name  \
0     January 15: Semi-Annual 2026     0.0        Report   Carolyn Arnold   
1     January 15: Semi-Annual 2026     0.0        Report     Zarin Gracey   
2     January 15: Semi-Annual 2026     0.0        Report   Jaime Resendez   
3     January 15: Semi-Annual 2026     0.0        Report  Monica R Alonzo   
4     January 15: Semi-Annual 2026     0.0        Report     Jesus Moreno   
...                            ...     ...           ...              ...   
8994     July 15: Semi-Annual 2024     0.0        Report    Kathy Stewart   
9002     July 15: Semi-Annual 2024     0.0        Report    Sandy Greyson   
9010     July 15: Semi-Annual 2024     0.0        Report  Cara Mendelsohn   
9014     July 15: Semi-Annual 2024     0.0        Report   Jaynie Schultz   
9022     July 15: Semi-Annual 2024     0.0        Report      Paul Ridley   

     Election Date    Transaction Date                   Street    City State  \
0              NaT 2026-01-15 17:54:31     835 Timber Dell Lane  Dallas    TX   
1              NaT 2026-01-15 16:48:27            PO Box 763173  Dallas    TX   
2       2025-05-03 2026-01-15 16:30:46    1458 Mission Hills Ln  Dallas    TX   
3       2025-05-03 2026-01-15 16:21:23              PO Box 4126  Dallas    TX   
4       2027-11-01 2026-01-15 16:21:19           4809 Victor St  Dallas    TX   
...            ...                 ...                      ...     ...   ...   
8994    2025-05-03 2024-07-12 19:07:58             PO Box 55081  Dallas    TX   
9002           NaT 2024-07-10 14:20:00         7238 Heathermore  Dallas    TX   
9010           NaT 2024-07-07 17:37:05      7120 Van Hook Drive  Dallas    TX   
9014    2025-05-03 2024-07-05 15:32:27  11222 Saint Michaels Dr  Dallas    TX   
9022           NaT 2024-07-02 09:39:06          5100 Victor St.  Dallas    TX   

     Zipcode                                       Geo Location  
0      75232  835 Timber Dell Lane Dallas TX 75232 (

In [78]:
zero_dollars = df_campfin[df_campfin['Amount'] == 0]
print(f"Number of entries with zero dollar amounts: {len(zero_dollars)}")

Number of entries with zero dollar amounts: 205


In [ ]:
# most of the $0 entries are candidate/committee contact type
zero_dollars['Contact Type'].value_counts()

Contact Type
Candidate / Committee    191
Expenditure               11
Report Itself              3
Name: count, dtype: int64

In [67]:
# check for duplicate rows
num_duplicates = df_campfin.duplicated().sum()
print(f"Number of duplicate rows: {num_duplicates}")

Number of duplicate rows: 0


In [68]:
# check distribution of key categorical variables
print(f"Unique candidates: {df_campfin['Candidate Name'].nunique()}")
print("\nRecord types:")
print(df_campfin['Record Type'].value_counts())
print("\nSchedule types:")
print(df_campfin['Schedule Type'].value_counts())

Unique candidates: 70

Record types:
Record Type
30th Day Before General Election 2025    2767
July 15: Semi-Annual 2025                1739
8th Day Before General Election 2025     1628
January 15: Semi-Annual 2025             1538
July 15: Semi-Annual 2024                 935
January 15: Semi-Annual 2026              713
8th Day Before Runoff Election 2025       552
Final Report (COH)                         78
Name: count, dtype: int64

Schedule types:
Schedule Type
Political Contributions Other Than Pledges Or Loans             5298
Political Expenditures                                          4016
Report Itself                                                    202
Report                                                           191
Political Expenditures Made From Personal Funds                  130
Loans                                                             55
Credits                                                           46
Notice From Political Committees           

### Exploration

In [69]:
# let's make a column that indicates whether the donor is a person or a business
def donor_type(row):
    if pd.notnull(row['Business Name']):
        return 'Business'
    elif pd.notnull(row['First Name']) and pd.notnull(row['Last Name']):
        return 'Individual'
    else:
        return 'Unknown'
    
df_campfin['Donor Type'] = df_campfin.apply(donor_type, axis=1)

In [70]:
# how many of each type?
df_campfin['Donor Type'].value_counts()

Donor Type
Individual    5682
Business      4268
Name: count, dtype: int64

In [71]:
# which candidates received the most in contributions?
df_campfin.groupby('Candidate Name')['Amount'].sum().sort_values(ascending=False).head(10)

Candidate Name
William Roth       1105927.16
Chad West           463768.06
Jeffrey Kitner      452369.70
Gay Willis          425329.09
Paula Blackmon      273083.51
Eric Johnson        263617.89
Lorie Blair         257646.49
Jesus Moreno        233520.80
Maxie Johnson       215220.22
Cara Mendelsohn     207487.82
Name: Amount, dtype: float64

In [72]:
# I want to look at out-of-state donors 
out_of_state_contributions = df_campfin[df_campfin['State'] != 'TX']
print(f"{len(out_of_state_contributions)} contributions are out of state")

945 contributions are out of state


In [73]:
out_of_state_contributions.sort_values('Amount', ascending=False).head()

ID  Record ID  Report ID  \
7616  346492      46492       2771   
3651  F48216      48216       2911   
3110  F48198      48198       2911   
1034  F48823      48823       2948   
1663  F48651      48651       2937   

                                              File Link First Name Last Name  \
7616  https://campfin.dallascityhall.com/FinalReport...        NaN       NaN   
3651  https://campfin.dallascityhall.com/FinalReport...        NaN       NaN   
3110  https://campfin.dallascityhall.com/FinalReport...        NaN       NaN   
1034  https://campfin.dallascityhall.com/FinalReport...        NaN       NaN   
1663  https://campfin.dallascityhall.com/FinalReport...        NaN       NaN   

               Business Name   Contact Type  \
7616       Charles Schwab Co  Report Itself   
3651            GDA Wins LLC    Expenditure   
3110            GDA Wins LLC    Expenditure   
1034  United States Treasury    Expenditure   
1663            GDA Wins LLC    Expenditure   

                               Record Type   Amount           Schedule Type  \
7616          January 15: Semi-Annual 2025  32037.0           Report Itself   
3651  8th Day Before General Election 2025  17264.7  Political Expenditures   
3110  8th Day Before General Election 2025  17264.7  Political Expenditures   
1034             July 15: Semi-Annual 2025   9906.0  Political Expenditures   
1663   8th Day Before Runoff Election 2025   9060.5  Political Expenditures   

      Candidate Name Election Date Transaction Date                    Street  \
7616     Paul Ridley    2025-05-03       2024-12-30              211 Main St.   
3651  Jeffrey Kitner    2025-05-03       2025-04-08             2001 S St. NW   
3110  Jeffrey Kitner    2025-05-03       2025-04-17             2001 S St. NW   
1034    Eric Johnson    2023-05-06       2025-06-02  1500 Pennsylvania Ave NW   
1663  Jeffrey Kitner    2025-06-07       2025-05-13             2001 S St. NW   

               City State Zipcode  \
7616  San Francisco    CA   94105   
3651     Washington    DC   20009   
3110     Washington    DC   20009   
1034     Washington    DC   20220   
1663     Washington    DC   20009   

                                           Geo Location Donor Type  
7616  211 Main St San Francisco CA 94105 (37.79033, ...   Business  
3651  2001 S St NW Washington DC 20009 (38.91409, -7...   Business  
3110  2001 S St NW Washington DC 20009 (38.91409, -7...   Business  
1034  1500 Pennsylvania Ave NW Washington DC 20220 (...   Business  
1663  2001 S St NW Washington DC 20009 (38.91409, -7...   Business

In [74]:
# Which candidates received the most out-of-state contributions?
out_of_state_contributions.groupby('Candidate Name')['Amount'].sum().sort_values(ascending=False).head()

Candidate Name
Jeffrey Kitner    112695.86
Eric Johnson       58127.85
Paul Ridley        45772.90
Paula Blackmon     37011.27
Marc Rossouw       27906.16
Name: Amount, dtype: float64